elastic/kibana
# [Fleet] Support browsing granular integrations

## Summary

Part of #93315. This PR has two main parts:

### Allow users to browse "granular integrations"
<details>
<summary>Screenshots</summary>

Granular AWS integration tiles on browse page
![image](https://user-images.githubusercontent.com/1965714/117894312-f156a200-b270-11eb-867c-19e88ecc8e8f.png)

Integration specific detail page
![image](https://user-images.githubusercontent.com/1965714/117895356-07656200-b273-11eb-8153-bf29f38870ef.png)

Installed integrations page, single AWS tile
![image](https://user-images.githubusercontent.com/1965714/117895447-38459700-b273-11eb-991e-47ee3cf6e31e.png)
</details>

- What this means: packages can now export multiple integrations by exporting multiple `policy_templates` in their package definition. Previously, all packages exported 1 policy template. Now, we have changes coming to the AWS package (https://github.com/elastic/integrations/pull/767) so that it exports multiple AWS integrations: AWS CloudWatch, AWS DynamoDB, etc.
- Each one of these integrations need to show as a separate icon/tile in Integrations > All integrations. The search functionality should work normally with these new tiles. Clicking one of these tiles will bring you to the integration detail page that shows *that integration's* specific icon, screenshots, and README content.
- In Integrations > Installed integrations, only the top-level package tile should show (aka only AWS tile) for now.
- All changes should be **BWC** with previous packages (aka packages that only export 1 policy template).


### Support package-level vars (user configurable variables)
<details>
<summary>Screenshots</summary>

Policy editor "Integrations settings" area expanded to include package-level vars
![image](https://user-images.githubusercontent.com/1965714/117895623-9bcfc480-b273-11eb-87bb-016577b7e6bd.png)

Package-level vars are used properly while compiling final agent YAML
![image](https://user-images.githubusercontent.com/1965714/117895692-c1f56480-b273-11eb-9b77-17532afbd450.png)
</details>

- Previously we only supported vars at the input- and stream- level. Now, packages can start using package-level vars. Fleet will use these package-level vars when compiling the final agent YAML.

## How to test
TLDR Build in-flight AWS package and load it into your local package registry

1. Get `elastic-package` https://github.com/elastic/elastic-package
2. Pull down https://github.com/elastic/integrations/pull/767
3. From that PR, cd into `<integrations>packages/aws`
4. Run `elastic-package build`, make note of the built `aws/0.6.0` package at `<integrations>/build`
5. Pull down and build `package-registry` https://github.com/elastic/package-registry
6. Copy the built `0.6.0` package to `<package-registry>/build/package-storage/packages/aws`
7. Run the package registry `go run .`
8. Edit `kibana.dev.yml` to use your local package registry by adding: `xpack.fleet.registryUrl: http://localhost:8080`
9. Start Kibana and observe the above changes to Integrations UI and policy editor. Play around to make sure I didn't break any other packages

## What's next

You'll notice that the policy editor for the AWS package only displays "Collect billing metrics". This is expected for now. The second part of the work to fully support all the requirements in #93315 is to adjust the policy editor to support multiple `policy_templates` too. Originally I tried putting everything in one PR but that second part is a bit more complex and it was holding up the rest of the work from going in 🙃
